### PR TITLE
Add entropy loss

### PR DESCRIPTION
**Pull request by pbokc**
Add entropy loss to address #583 

Ran FSDP with and without entropy loss with Qwen2.5-0.5B-Instruct and saw that policy entropy remained ~stable with entropy loss off and it increased rapidly with entropy loss on with a coefficient of 10. 
W B Chart 11_5_2025, 8_18_42 PM
W B Chart 11_5_2025, 8_19_56 PM

### CODE REVIEW

File: skyrl-train/skyrl_train/config/ppo_base_config.yaml [hunk 1]
```diff
@@ -90,6 +90,9 @@ trainer:
     use_kl_in_reward: false # apply kl loss to rewards
     use_kl_loss: true # used in policy model
     kl_loss_coef: 0.001
+    # entropy loss (disable by default?)
```
Comment (pbokc) [inline skyrl-train/skyrl_train/config/ppo_base_config.yaml]: Disable by default for backward compatibility?

File: skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 2]
```diff
@@ -233,10 +233,15 @@ def loss_func(logits, data):
...
 
-            with torch.no_grad():
+            with torch.set_grad_enabled(self.cfg.trainer.algorithm.use_entropy_loss):
                 action_logits = logits[:, -num_actions - 1 : -1, :]
                 entropy_BS = vocab_parallel_entropy(action_logits)
-                entropy = entropy_BS.sum().item() / entropy_BS.numel()
+                entropy_scalar = masked_mean(entropy_BS, loss_mask)
+
+            if self.cfg.trainer.algorithm.use_entropy_loss:
+                policy_loss = policy_loss - self.cfg.trainer.algorithm.entropy_loss_coef * entropy_scalar
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py]: Can you make this fix @pbokc

[the remark this answers is not in the document]
Comment (pbokc) [reply skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py]: updated

File: skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 3]
```diff
@@ -248,12 +248,13 @@ def loss_func(logits, data):
                 kl_loss = masked_mean(kl_loss, loss_mask, dim=-1).mean()
             else:
                 kl_loss = torch.tensor(0.0)
+            kl_loss_term = kl_loss * self.cfg.trainer.algorithm.kl_loss_coef
 
-            loss = policy_loss + kl_loss * self.cfg.trainer.algorithm.kl_loss_coef
+            loss = policy_loss + kl_loss_term - entropy_loss_term
```
[the remark this answers is not in the document]
Comment (pbokc) [reply skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py:258]: if `self.cfg.trainer.algorithm.use_entropy_loss` is `False`, `entropy_loss_term` will be set to 0. So subtracting it will do nothing.

File: skyrl-train/skyrl_train/config/ppo_base_config.yaml (hunk 1 above)
Comment (SumanthRH) [reply skyrl-train/skyrl_train/config/ppo_base_config.yaml]: Yes, makes sense

File: skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 4]
```diff
@@ -233,10 +233,15 @@ def loss_func(logits, data):
...
+            with torch.set_grad_enabled(self.cfg.trainer.algorithm.use_entropy_loss):
                 action_logits = logits[:, -num_actions - 1 : -1, :]
                 entropy_BS = vocab_parallel_entropy(action_logits)
-                entropy = entropy_BS.sum().item() / entropy_BS.numel()
+                entropy = masked_mean(entropy_BS, loss_mask)
+
+            if self.cfg.trainer.algorithm.use_entropy_loss:
+                entropy_loss_term = entropy * self.cfg.trainer.algorithm.entropy_loss_coef
+            else:
+                entropy_loss_term = torch.tensor(0.0)
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl-train/skyrl_train/workers/megatron/megatron_model_wrapper.py:244]:
```python
Python 3.12.11 | packaged by conda-forge | (main, Jun  4 2025, 14:45:31) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> a = torch.tensor(0.0)
>>> a.to("cuda") + a
tensor(0., device='cuda:0')
>>> 
```

@gemini-code-assist

**Comment by SumanthRH** [comment]
Thanks @pbokc ! I made some updates to include this in the tests for FSDP and Megatron backends and also added the final loss (after aggregation and normalization with accumulation steps) as `final_loss`
